### PR TITLE
Octoprint vcgencmd patch

### LINKED ISSUE #210629: octoprint-pisupport: Cannot run `vcgencmd` to get throttle status
### Describe the bug

[`octoprint-pisupport`](https://github.com/NixOS/nixpkgs/blob/89e49d8/pkgs/applications/misc/octoprint/default.nix#L53-L71) is a bundled OctoPrint plugin that fetches diagnostic information when running on a Raspberry Pi.

The plugin tries to run `vcgencmd`, a utility included in `libraspberrypi`. The default path in the plugin is `/usr/bin/vcgencmd`, which does not exist on NixOS, so the command fails.

Fixing that is easy enough, either by patching `octoprint-pisupport` or by setting the `vcgencmd_throttle_check_command` plugin config key. However, the command still fails:

```
octoprint.plugins.pi_support - WARNING - Cannot parse "/nix/store/6rc4gk09hd1cq3h4l8xll3nyzwl7sm7i-libraspberrypi-unstable-2022-03-23/bin/vcgencmd get_throttled" output: VCHI initialization failed
```

OctoPrint then recommends that you add the `octoprint` user to the `video` group, but that doesn't fix the problem. The issue is that `vcgencmd` is trying to access `/dev/vchiq`, the interface to the video controller on the Pi. On Raspberry Pi OS, `/dev/vchiq` is owned by `root:video` with `660` permissions. On NixOS, it's `root:root` with 600.

I tried setting `DeviceAllow=/dev/vchiq rw` in `octoprint.service`, but that didn't work either. I am not sure why, or how to diagnose that issue further.

### Steps To Reproduce

1. Configure `services.octoprint.enable = true` on a Raspberry Pi running NixOS
2. `journalctl -u octoprint.service | grep pi_support`

### Expected behavior

The plugin should find `vcgencmd` and successfully run it with no additional configuration required by the user.

### Notify maintainers

@abbradar @gebner @WhittlesJr @gador

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"aarch64-linux"`
 - host os: `Linux 5.15.74, NixOS, 22.11 (Raccoon), 22.11.20221224.dirty`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.1`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`
```

### PR DESCRIPTION
**Pull request by gador**
###### Description of changes

`vcgencmd` is used on raspberry pi in octoprint to check for throttling. Until now, it didn't work, because the path is hardcoded and the permissions of the device `/dev/vchiq` did not allw non-root users any access.

~~This PR introduces the corresponding `udev` rule to allow the group `dialout` access to the device and patch the hardcoded path.~~

EDIT: It introduces ACL permission setting of `dev/vchiq` to give the user `octoprint` `rw` permission on that device. This should mitigate attack surface in contrast to giving a group access to that device. (as per @NickCao suggestion)

This fixes #210629. 

Tested on a aarch64 raspberry pi 4 to work correctly.

~~Please note: PR #213145 is also included, since `octoprint` won't start without it.~~

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/octoprint/default.nix [hunk 1]
```diff
@@ -82,6 +82,12 @@ let
 
               # requires octoprint itself during tests
               doCheck = false;
+              patches =  [
+              (pkgs.substituteAll {
+                src = ./vgencmd.patch;
+                vcgencmd = "${self.pkgs.libraspberrypi}/bin/vcgencmd";
+              })
+            ];
```
Comment (NickCao) [inline pkgs/applications/misc/octoprint/default.nix:90]: Can't we just do
```
  substituteInPlace octoprint_pi_support/__init__.py --replace /usr/bin/vcgencmd ${self.pkgs.libraspberrypi}/bin/vcgencmd
```
in postPatch

Comment (gador) [reply pkgs/applications/misc/octoprint/default.nix:90]: It should theoretically work like that, too. Shall I change it?

Comment (NickCao) [reply pkgs/applications/misc/octoprint/default.nix:90]: That would be cleaner to read and easier to maintain.

File: pkgs/applications/misc/octoprint/default.nix [hunk 2]
```diff
@@ -209,6 +215,15 @@ let
                 rm pytest.ini
               '';
 
+              udevRules = ''
+                # Change /dev/vchiq from root:root 0600 to root:dialout  0660 for vcgencmd to work
+                SUBSYSTEM=="vchiq", GROUP="dialout", MODE="0660"
```
Comment (NickCao) [inline pkgs/applications/misc/octoprint/default.nix:220]: The commit adding dialout to SupplementaryGroups https://github.com/NixOS/nixpkgs/commit/f8df7606683fc9fbce68c8c363d1318dab2ed199 says dailout is for serial devices. I do lack the domain knowledge about this, what is vchiq?

File: pkgs/applications/misc/octoprint/default.nix (hunk 1 above)
Comment (gador) [reply pkgs/applications/misc/octoprint/default.nix]: ok

File: pkgs/applications/misc/octoprint/default.nix (hunk 2 above)
Comment (NickCao) [reply pkgs/applications/misc/octoprint/default.nix:218]: https://raspberrypi.stackexchange.com/a/54590
The post suggests that vchiq is a decently dangerous interface to expose to random programs, hence the somewhat restrictive permissions by default. Not sure if we really want to loosen that permission.

Comment (gador) [reply pkgs/applications/misc/octoprint/default.nix:218]: It is a raspberry pi specific program to check internals about the hardware (like throttling). On native devices running something like ubuntu etc it seems to be owned by the `video` group with Mode 0660.
I decided against changing the group to `video` and adding `octoprint` to that group so to give the least privileges to the service, while still maintaining the function to run it.

Comment (gador) [reply pkgs/applications/misc/octoprint/default.nix:218]: Well it seems some people set it to Mode `777` for programs to `just work`.
I think `dialout` is a specific enough group to get access to that device. But I'm open to suggestions to change that

Comment (NickCao) [reply pkgs/applications/misc/octoprint/default.nix:218]: Can we use ACL to grant only `octoprint` access to vchiq. `tmpfiles.d` can be used to set ACL on files.

File: pkgs/applications/misc/octoprint/default.nix [hunk 3]
```diff
@@ -209,6 +213,7 @@ let
                 rm pytest.ini
               '';
 
+
```
Comment (NickCao) [inline pkgs/applications/misc/octoprint/default.nix:216]: Remove the newline here.

File: nixos/modules/services/misc/octoprint.nix [hunk 4]
```diff
@@ -106,6 +106,8 @@ in
 
     systemd.tmpfiles.rules = [
       "d '${cfg.stateDir}' - ${cfg.user} ${cfg.group} - -"
+      # this will allow octoprint read access to raspberry specific hardware to check for throttling
+      "a /dev/vchiq - - - - u:octoprint:r"
```
Comment (NickCao) [inline nixos/modules/services/misc/octoprint.nix:110]: Please test if it works on a rpi and does not cause failures on what's not a rpi (where there is no /dev/vchiq), especially the status of `systemd-tmpfiles-setup.service`.

**Review by NickCao — APPROVED** [review]

File: nixos/modules/services/misc/octoprint.nix (hunk 4 above)
Comment (gador) [reply nixos/modules/services/misc/octoprint.nix:110]: testing right now

Comment (gador) [reply nixos/modules/services/misc/octoprint.nix]: My testing showed:

- No errors on x86-linux systems (tested in a VM) where no `dev/vchiq` exists. `systemctl status systemd-tmpfiles-setup.service` does not show any errors, just like `journalctl -xeu systemd-tmpfiles-setup.service`
- No errors show on raspberry pi 4 (just like above)
- correct permissions are added as confirmed by `getfacl /dev/vchiq` 
- `octoprint` can use `vcgencmd` and confirmed per CLI with `sudo -u octoprint vcgencmd get_throttled`
- `octoprint` WebUI doesn't show `vcgencmd` error anymore and seems to detect correct sate (as seen in startup logs)

File: pkgs/applications/misc/octoprint/default.nix (hunk 3 above)
Comment (NickCao) [inline pkgs/applications/misc/octoprint/default.nix:216]:
```suggestion
```

**Pull request merged**